netstacklat - Handle TCP head-of-line blocking

simosund opened the pull request:
Currently netstacklat may report latency for head-of-line (HOL) blocked TCP reads. TCP HOL-blocking can occur when there is packet loss or reordering, when the TCP stack has to delay the delivery of a packet to the socket receive queue until the preceding TCP segments have been received.

As this HOL-blocking is dependent on factors external to the local host (i.e. the packet loss and latency of the entire network path from the sender), this is not the type of latency we typically want to highlight with netstacklat. This PR addresses this issue by not recording the latency for reads that may experience HOL-blocking.

See the commit messages for further details on the how this is accomplished and some limitations.

simosund committed:
Minor refactor of network filtering

The filter_ifindex() and filter_network_ns() functios are often used
together. Therefore, add a filter_network() function that budles these
two together.

Signed-off-by: Simon Sundberg <simon.sundberg@kau.se>

simosund committed:
netstacklat: Exclude TCP reads for HOL blocked segments

The 'tcp-socket-read' currently reports the latency for the skb
containing the last TCP segment read from the socket. However, this
segment might have been head of line (HOL) blocked by a previous
segment missing. In this case, netstacklat's reported latency will
include HOL blocking periods that is dependent on external
factors (such as network packet loss, and network latency impacts
retransmission time). As netstacklat is primarily intended to identify
issues within the local host (in the network stack or receiving
applications), by default filter out any socket reads were the last
read SKB might have experienced HOL-blocking.

Add the new -y/--include-tcp-hol-delay option to retain the old
behavior of reporting latency for all reads, including those that are
HOL-blocked. This may be useful in some scenarios when you still want
to be aware of latency issues caused by HOL-blocking, even though it
is caused by external components. For example, in a data center
context were you have full control over the network, it may still be
relevant to monitor HOL-based caused by the network.

To exclude HOL-blocked reads, detect if any new ooo-segments have
arrived by checking for differences in the number of ooo-packets in
tcp_sock->rcv_ooopack. If any new ooo-segments have arrived, exclude
the latency sample from the current read and set a limit for the next
safe sequence number to read where the current ooo-packets must have
been passed so segments can no longer be HOL-blocked. If there are
skbs in the ooo-queue, set the limit to the end of the
ooo-queue. Otherwise, set the limit to the current rcv_nxt (as if the
ooo-queue is empty the detected ooo-segments must already have been
merged into the receive queue and rcv_nxt must have advanced past
them). If the read is past the safe sequence limit and no new
ooo-segments have arrived, it's safe to start including the latency
samples again.

For sockets were some ooo-segments have been observed, keep the
ooo-range state in socket storage (BPF_MAP_TYPE_SK_STORAGE). Skip
protecting this state with a spin-lock, as it should only be
concurrently accessed if there are concurrent reads on the same TCP
socket, which is assumed to be very rare as applications attempting
that cannot know which part of the data each of their concurrent reads
will get.

There are some scenarios that may cause this ooo-filtering to fail.
- If multiple reads are done to the socket concurrently, we may not
  correctly track the last read byte. The kernel does not keep a lock
  on the TCP socket at the time our hooked function
  tcp_recv_timestamp() runs. If two reads are done in parallel, it's
  therefore possible that for both reads we will check the last read
  byte (tcp_sock.copied_seq) after the second read has updated it. We
  may then incorrectly conclude that the first read was ahead of the
  ooo-range when it was not, and record its latency when we should
  have excluded it. In practice I belive this issue should be quite
  rare, as most applications will probably not attempt to perform
  multiple concurrent reads to a single connected TCP socket in
  parallel (as then you cannot know which part of the payload the
  parallel reads will return).

- As tcp_recv_timestamp() runs outside of the socket lock, the various
  state members we access may concurrently be updated as we're
  attempting to read them. An especially problematic one is
  tcp_sock.ooo_last_skb, which keeps a pointer to an SKB that is only
  valid while the ooo-queue is non-empty. It is possible that between
  our check for if the ooo-queue is non-empty and following the
  ooo_last_skb pointer, the ooo-queue is cleared and the ooo_last_skb
  pointer may end up pointing towards a freed SKB. If the socket
  members we access are updated before or while we read them, it can
  break the filtering in numerous ways, e.g. result in includes
  samples that should have been excluded (due to e.g. copied_seq being
  updated before our read) or excluding a large amount of valid
  samples (due to e.g. setting a sequence limit based on garbage in a
  freed SKB).

Signed-off-by: Simon Sundberg <simon.sundberg@kau.se>

simosund commented:
@netoptimizer In my last commit I've added some sanity checks like we talked about on the meeting. I am however not sure if these sanity checks are a very good idea, as they do come with a noticeable increase in overhead and still offer somewhat limited protection.

### Performance impact
I've rerun my test with various degrees of packet loss, and the results are available in [this dashboard](https://snapshots.raintank.io/dashboard/snapshot/zKw4xyfSLQSPRGLppWRJ0xnW6UdkCbHI). The first run is with current master (no ooo-filter at all), the second with "alt3" as of commit dff6bc54988ea (without sanity checks, same as alt3 in my previous test), and the last is alt3 with the sanity checks added (in commit 59e923df588d6).

Looking at performance for the tcp-socket-read hook (the only one that should be affected), it looks like this:

#### Program runtime
<img width="1434" height="727" alt="retrans_checksoverhead_runtime" src="https://github.com/user-attachments/assets/8e5cdef5-08c1-4b17-a4d1-dd4f1325fadc" />

*NOTE:* The spikes to ~700ns here are from periods of very little traffic, so focus on the parts between the spikes.

Runtime without ooo-filter varies from around 250 - 290ns, runtime with ooo-filter varies from around 300 - 390ns, and runtime with ooo-filter + added sanity checks varies from around 350 - 470ns. So the sanity checks add about as much additional overhead as the entire filtering mechanism itself. Or phrased a differently, the sanity checks roughly double the overhead of the entire filtering mechanism as a whole.

#### Overall CPU util
<img width="1434" height="727" alt="retrans_checksoverhead_cpu" src="https://github.com/user-attachments/assets/966adf75-4170-407e-9d44-de36be10c0eb" />

The largest CPU utilization overall seems to occur at the test with 0.1% packet loss (second spike in each group). For this test the CPU utilization without ooo-filtering is about 0.34%, about 0.46% with ooo-filtering and around 0.57% with ooo-filtering + sanity checks. So once again, the sanity checks roughly double the overhead of the ooo-filtering.

### Limitations with the sanity checks
The sanity checks are described in the commit message (and the code). There's basically two types of them.
- A sanity check that the sequence number (`skb->cb.end_seq`) read from `tp->ooo_last_skb` is within the current receive window
- Sanity checks that `tp->rcv_nxt` and `tp->copied_seq` did not decrease since last read

The sanity check that the ooo sequence number is within the receive window adds some overhead as it requires probing the tp->rcv_wup and tp->rcv_wnd fields as well. It also adds a potential new error source as now we need to trust that they are right, although it's less likely they're catastrophically wrong than the skb read (as they don't risk being read from a freed skb). 

I'm also not entirely sure that the end of the ooo-sequence number must always be within the current receive window, even if that should usually be the case. Partly because depending on system configuration it appears that the receive window can shrink (see the [tcp_select_window()](https://elixir.bootlin.com/linux/v6.17.4/source/net/ipv4/tcp_output.c#L279)), in which the current receive window might actually be smaller than the receive window was when the skb was added to the ooo-queue. And partly because as far as I can tell from the code in [tcp_data_queue()](https://elixir.bootlin.com/linux/v6.17.4/source/net/ipv4/tcp_input.c#L5155) it only drops packets where the starting sequence (cb.seq) is ahead of the receive window, meaning that segments where the end sequence (cb.end_seq) is ahead of the receive window may still be added to the error queue.

I'm also not sure how much the "doesn't decrease between reads" checks on rcv_nxt and copied_seq really help, as even if we get a complete garbage value (which we generally shouldn't here though, as the socket should still be valid while our eBPF program runs and 32-bit writes I assume are "atomic" on most architectures) there's only about a 50% chance we will detect the error. Or put another way, it won't detect errors below ~2GB. It might be possible to reel in these limits further if there's some good way to determine the maximum amount of data that can be data that can be read in a single TCP socket read, but even if possible that will likely require probe reading additional socket members (adding additional overhead and error sources).

### Conclusion
The sanity check on the data in the ooo_last_skb would be really good to have (as that is read from a potentially invalid pointer), but it does add some overhead (and new potential error sources) due to having to probe additional fields, and I'm not sure the concept behind it (comparing it against the current receive window) is entirely sound.

The sanity checks on `rcv_nxt` and `coped_seq` are quite weak. They also add some overhead as now they always have to be probed to keep track of their state, while before they only needed to be probed conditionally when needed.

simosund commented:
Hi again @netoptimizer,
I've now split up these sanity checks into two different commits. The first one adds the sanity check for `tp->ooo_last_skb->cb.end_seq` value (that it's non-zero and within the current receive window), while the second commit adds the monotonic growth checks for `tp->rcv_nxt` and `tp->copied_seq`. TLDR, I recommend including the first these commits, and skipping the second.

I think it makes sense to adapt the first of these sanity checks, as that's the only one were we have to follow an additional pointer to a potentially no longer valid SKB. As I mentioned previously, I think there might be some circumstances where `end_seq` of the ooo-queue can actually end up ahead of the current receive window, so this check might be a bit too strict. In some cases the receive window can shrink, but only if `net.ipv4.tcp_shrink_window` is set (1). However, the default is for it to be unset (0) and I doubt it's particularly common that people change it (although please check its value on your systems in case it's more common than I think). As I also noted previously, as far as I can tell SKBs where the `end_seq` but not `seq` (i.e. the start seq) is ahead of receive window may be added to the ooo-queue. However, after also going through the code in `tcp_rcv_established()` (which is the one that calls `tcp_data_queue()`, which in turns calls `tcp_data_queue_ofo()`) I saw there were some additional checks that ultimately culminates in `tcp_sequence()`, where an SKB with an `end_seq` that overruns the current receive window is only accepted if the receive queue is currently empty. So requires very specific conditions for that to go wrong (ooo-segment where start-seq is within the receive window but end-seq is beyond it arrives while the receive queue is empty). 

I'm more skeptical to the sanity checks on `rcv_nxt` and `copied_seq`. As I've previously mentioned they're rather weak checks, and require that we track and update their state every time. They're also for values that are direct members of the tcp_sock, so the values we get for them should still be valid (although possibly updated before or in-between we read them), as long as only a part of the value is updated at a time (e.g. the first/last 16 bits). But 32-bit writes should effectively be "atomic" (as it either you'll read the full value before an update, or the full value after an update) on most architectures, right?

## Performance overhead
Through the test below I also show that the the `tp->ooo_last_skb->cb.end_seq` sanity check seems to have neglible additional overhead compared to alt3 ooo-filtering without any sanity checks. While also adding in the sanity checks for `rcv_nxt` and `copied_seq` roughly doubles the overhead of the ooo-filtering. 

I've done an additional set of tests were I run:
- no ooo-filtering (current master)
- alt3 ooo-filtering without any sanity checks ()
- alt3 ooo-filtering with `tp->ooo_last_skb->cb.end_seq` sanity check ()
- alt3 ooo-filtering with both `tp->ooo_last_skb->cb.end_seq` + `rcv_nxt` and `copied_seq` sanity checks

As usual, full dashboard with interactive (mouseover) plots are available [here](https://snapshots.raintank.io/dashboard/snapshot/YuyP8OiUotmywX3CyxG0TX8mV8XNYCCS). For convenience, I've inserted the performance-related figures below:

<img width="1134" height="622" alt="Screenshot 2025-10-27 at 18-42-03 View panel - netstacklat - TCP HoL blocking - none alt3 alt3 ooo-seq-check alt3 ooo-seq-check rcv_nxt_copied-seq-check - Grafana" src="https://github.com/user-attachments/assets/999e4f7a-3231-4e91-8436-4d0c3cd4e77c" />

<img width="1134" height="622" alt="Screenshot 2025-10-27 at 18-43-27 View panel - netstacklat - TCP HoL blocking - none alt3 alt3 ooo-seq-check alt3 ooo-seq-check rcv_nxt_copied-seq-check - Grafana" src="https://github.com/user-attachments/assets/8ad59f04-483d-485a-83a3-c367151dce09" />

simosund commented:
Right, forgot that this hadn't been merged yet. I don't think anyone has formally reviewed it, but I assume @netoptimizer in practice ended up reviewing it while applying it to the Cloudflare build?

As I mention in the discussion above, I also think we should skip the last commit, as those sanity checks add considerable overhead while it's doubtful they would really catch many real issues. So unless anyone objects about that, I should drop that before we merge.

tohojo commented:
I'm OK with dropping the last commit if it doesn't matter in practice. Ping @netoptimizer re: the review status? :)

tohojo commented:
Any updates on this, @simosund ? As I recall, you were planning to drop the last commit from the series?

simosund commented:
Yes, I was initially waiting on @netoptimizer to go through it, but I'm fairly confident that he never applied the last commit either, and in my own evaluation I did not use the last commit. So let's just drop it.

From my understanding the plan was for @netoptimizer to review it, but as he's currently on vacation I guess the status there is unchanged.

tohojo commented on netstacklat/netstacklat.bpf.c:
```diff
@@ -300,12 +330,156 @@ static bool filter_min_sockqueue_len(struct sock *sk)
+		window = rcv_nxt;
+
+exit:
+	*seq = window;
```
nit: the usual pattern is to only write to the return variable if there is no error.

simosund replied:
Fixed. As there's nothing that needs to be done on errors now, I've replaced the goto exit with early returns.

tohojo commented on netstacklat/netstacklat.bpf.c:
```diff
@@ -300,12 +330,156 @@ static bool filter_min_sockqueue_len(struct sock *sk)
+	}
+
+exit:
+	*seq = max_seq;
```
Same here

simosund replied:
Fixed. As there's nothing that needs to be done on errors now, I've replaced the goto exit with early returns. Furthermore, seq is now set directly instead of using the intermediate max_seq variable.

tohojo reviewed (changes requested):
A few nits, otherwise I think we should just merge this :)

simosund committed:
netstacklat: Add sanity check for out-of-order sequence

The logic for excluding samples from TCP reads that may have been
delayed by HOL blocking relies on reading a number of fields from the
TCP socket outside of the socket lock. This may be prone to errors due
to the socket state being updated at another place in the kernel while
our eBPF program is running. To reduce the risk that a data race
causes the filter to fail, add a sanity check for the maximum out of
order sequence used to exclude future TCP reads from monitoring.

The most problematic of the read fields in the tcp_sock is
ooo_last_skb, as that is a pointer to another SKB rather than a direct
value. This pointer is only valid as long as the out_of_order_queue is
non-empty. Due to a data race, we may check that the ooo-queue is
non-empty while there are still SKBs in it, then have the kernel clear
out the ooo-queue, and finally attempt to read the ooo_last_skb
pointer later when it is no longer valid (and may now point to a
freed/recycled SKB). This may result in incorrect values being used
for the sequence limit used to exclude future reads of
ooo-segments. The faulty sequence limit may both cause reads of
HOL-blocked segments to be included or the exclusion of an
unnecessarily large amount of future reads (up to 2 GB).

To reduce the risk that the garbage data from an invalid SKB is used,
introduce two sanity checks for end_seq in the ooo_last_skb. First
check if the sequence number is zero, if so assume it is invalid (even
though it can be a valid sequence number). Even though we will get an
error code if reading the data from this SKB fails altogether, we may
still succeed reading from a no longer valid SKB, in which case there
is a high risk the data will have been zeroed. If it's non-zero, also
check that it is within the current receive window (if not, clamp it
to the receive window).

Signed-off-by: Simon Sundberg <simon.sundberg@kau.se>

tohojo reviewed (approved):
Alright, let's merge this then :)
